created atomic transactions chain hip draft (#551)
hashgraph/hedera-improvement-proposal: HIP/hip-551.md

---
hip: 551
title: Batch transactions
author: Piotr Swierzy <piotr.swierzy@arianelabs.com>
working-group: Nick Poorman <@nickpoorman>, Michael Garber <@mgarbs>, Ashe Oro <@Ashe-Oro>, Bug Bytes <@bugbytes>, Patrick Woo-Sam <@pwoosam>
type: Standards Track
category: Service
needs-council-approval: Yes
status: Review
created: 2022-07-25
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/531
updated: 2023-10-05
---

## Abstract

This HIP defines a mechanism to execute batch transactions such that a series of transactions(HAPI calls) depending on each other can be rolled into one transaction that passes the ACID test (atomicity, consistency, isolation, and durability)

## Motivation

The existing implementation of transactions in the Hedera network does not allow multiple different HAPI transactions to be called in one single network transaction that would have all the ACID properties.
This makes it impossible to create more complicated flows without using smart contracts (which do not support all the HAPI transactions at this point) and listening to the mirror node to check the status of the previous transaction.

> This way we can also achieve an abstraction away from smart contracts

Reviewer: ```suggestion
This way we can also achieve an abstraction away from smart contracts.
```

That way a backend developer could create complicated flows without the need to use smart contracts and the need to handle partially successful flows in the backend,
e.g. the wipe transaction was successful but the mint transaction was not.

4. As a Hedera service user, I want to be able to create batch transaction, that use multiple Hedera services e. g.
I want to be able to wipe a token, mint a token, transfer a token, and submit a topic message in one ACID transaction.
That way a backend developer could create complicated flows without the need to use smart contracts and the need to handle partially successful flows in the backend,
e.g. the wipe transaction was successful but the mint transaction was not.

5. As a Hedera token admin, I want to be able to gift Alice some tokens. I want to be able to mint and transfer the tokens in a single transaction.

6. As a Hedera token owner, I want to be able to wipe tokens from an account that I revoked the KYC. I want to be able to grant KYC, wipe tokens and revoke KYC in a single transaction.

## Specification

### Transaction Integrity

> As emphasized by `Rbair23`:

Reviewer: ```suggestion
```

1. Each inner transaction within a batch should be treated as a self-contained transaction, unaware of its batch context.
2. If an inner transaction fails, all preceding inner transactions that succeeded still have to pay, even if their effects are not committed.

> 3. Inner transactions are regular transactions, and the batch is signed by all involved parties.

Reviewer: ```suggestion
3. Inner transactions are individually signed regular transactions indicating that once a person signs an inner transaction, they have agreed to its outcome.
```

> 4. Inner transactions must remain secure, preventing man-in-the-middle attacks and batch tampering. This preserves the intent of the transaction and ensures the sender's intentions are clear.

Reviewer: ```suggestion
```

> However, `se7enarianelabs` counters the necessity for every inner transaction signer to also sign the batch, indicating that once a person signs an inner transaction, they have agreed to its outcome. This feedback suggests that the context of a transaction shouldn't overrule the transaction's content.

Reviewer: ```suggestion
```

### Batch Signatures

Each transaction must be individually signed. Signatures on the batch itself are unique to the batch and don't represent the inner transactions within. This way:

- It's impossible for inner transactions to be rearranged post-signature.

> - Man-in-the-middle and malicious node activities, such as front-running, remain a concern. Implementing TLS connections to nodes and choosing trusted nodes can counteract this.

Reviewer: ```suggestion
- Batch transactions are not designed to solve a problem for retail users but rather for developers and dApps. When someone signs a transaction, they individually agree to the outcome, thus the individual transaction signers within the batch need not sign the batch itself.
```

  
### Handling Inner Transactions

Inner transactions, by design:

1. Are signed individually.
2. Each pays for itself (node + network + service fees).
3. Have their own payer.
4. Are deduplicated on an individual basis.
5. Are authorized solely by their keys; batch keys are not used.

The batch:

1. Has its own payer and signatures.
2. Considers inner transactions as part of its signed bytes.
3. Makes the batch payer cover node+network fees for handling the batch, excluding inner transaction fees.
4. Has deduplication based on its unique transaction ID.

> We could create a new protobuf message based on `SignedTransaction` message:

Reviewer: ```suggestion
5. The submitter does not need to be authorized to carry out all the inner batch transactions. As with every transaction in Hedera, where the submitter can be any account as long as the transaction is properly signed.
6. The standard max transaction size must be respected.

We could create a new protobuf message based on `SignedTransaction` message:
```

---
hip: 551
title: Batch transactions
author: Piotr Swierzy <piotr.swierzy@arianelabs.com>
type: Standards Track
category: Service
needs-council-approval: Yes
status: Review
created: 2022-07-25
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/531
updated: 2022-09-19, 2022-10-04
---

## Abstract

This HIP defines a mechanism to execute batch transactions such that a series of transactions(HAPI calls) depending on each other can be rolled into one transaction that passes the ACID test (atomicity, consistency, isolation, and durability)

## Motivation

The existing implementation of transactions in the Hedera network does not allow multiple different HAPI transactions to be called in one single network transaction that would have all the ACID properties.
This makes it impossible to create more complicated flows without using smart contracts (which do not support all the HAPI transactions at this point) and listening to the mirror node to check the status of the previous transaction.

> This way we can also achieve an abstraction away from smart contracts

Reviewer 2: This also means that if any of the contract calls at the end are reverted, then the approvals, associates, and account updates would be undone. This would improve the security of the network by reducing dangling allowances and auto association slots.

## Rationale
Atomic transactions chain is from the protobuf point of view just a list of `SignedTransactions` that's, why we based our new protobuf message on it,
we only have to change the `bodyBytes` field to be `repeated`, that way we can store there multiple transaction bodies,
`SignatureMap sigMap` can stay unchanged because it's defined as `repeated SignaturePair sigPair`.

## User stories

> As a Hedera token service user, I want to be able to unfreeze an account, send an NFT, and freeze it again in one ACID transaction, that way I can achieve an account-bound NFT(nontransferable NFT) collection, without using the hedera smart contract service.

Reviewer 3: nit: should number the list
```suggestion

1. As a Hedera token service user, I want to be able to unfreeze an account, send an NFT, and freeze it again in one ACID transaction, that way I can achieve an account-bound NFT(nontransferable NFT) collection, without using the hedera smart contract service.

```

We can use nontransferable FT e.g. as game points, reputation points, etc.

As a Hedera token service user, I want to be able to wipe a token, mint a token, and transfer a token in one ACID transaction.
That way a backend developer could create complicated flows without the need to use smart contracts and the need to handle partially successful flows in the backend,
e.g. the wipe transaction was successful but the mint transaction was not.

As a Hedera service user, I want to be able to create batch transaction, that use multiple Hedera services e. g.
I want to be able to wipe a token, mint a token, transfer a token, and submit a topic message in one ACID transaction.
That way a backend developer could create complicated flows without the need to use smart contracts and the need to handle partially successful flows in the backend,
e.g. the wipe transaction was successful but the mint transaction was not.

## Specification
We could create a new protobuf message based on `SignedTransaction` message:
```

> message BatchTransaction {

Reviewer 3: Q: Is it fair to say the assumption or requirement is that the submitter is authorized to carry out all the batch transactions?
For example to carry out some of the example one be both the treasurer, in another case the freezeKey owner.

Author: The submitter does not need to be authorized to carry out all the batch transaction, every transaction in the batch needs to be signed by all parties involved in the individual transaction.

It would work like every transaction in Hedera, where the submitter can be any account as long as the transaction is properly signed

    /**
    * TransactionBodies serialized into bytes, which must be signed
    */

>     repeated bytes bodyBytes = 1;

Reviewer 3: Q: do each of these transactions result in externalized individual transaction records in the record stream?
If individual records are exposed  is there a concept of parent and child transactions? If so you should supply an example of this.

Reviewer 4: Instead, it might be possible to modify the `TransactionBody` protobuf message with a new field inside the `data` field that would be a repeated field of `TransactionBodyXXXX`  (now this is where it gets murky, should it be a repeated field of encoded transaction body XXX,  or should it be an explicitly defined new message that is a union of all the possible TransactionBodyXXXX types)  Either way, with an ordered list of TransactionBodyXXXX you can achieve the desired outcome and only have to generate the enclosing Transaction ID for the outer TransactionBody envelope.

>     repeated bytes bodyBytes = 1;

Reviewer 3: Q: Also how should the network manage when one of these transaction has a child transaction?

Author: If the parent transaction is in the batch and the batch isn't successful, then the child transaction should also be reverted.

Reviewer 5: that's how i see it.  All or nothing.

>     repeated bytes bodyBytes = 1;

Reviewer 3: Q: Are there any applicable byte size limits for consideration by the SDK and the network itself?

Author: I think, the standard transaction size limit should be respected

    /**
     * The signatures on the body with the new format, to authorize the transactions
     */
    SignatureMap sigMap = 2;
}
```
We could also create a predefined set of batch transactions and add new elements to it in the future
if the community asks for it.<br>
Initial set:
```

> Unfreeze -> Transfer -> Freeze

Reviewer 3: Are there any non HTS cases ? Is this for HTS only or all HAPI transactions?
If there's a set on non applicable transactions those should be called out

Reviewer 6: It would be more practical if these were for all HAPI transactions. Any application that needs a hapi transaction to succeed in order to do something else would use this. Gated topics or consensus messages have huge potential for content creation use cases. Here are a couple of scenarios:
Transfer -> Topic Create -> Submit Topic Message
Transfer -> Submit Topic Message

Reviewer 7: The ability to combine HTS with HCS would be super powerful if they are batched. So, would also agree for all HAPI transactions to reach the maximum value out of this HIP proposal.

Wipe -> Mint -> Transfer
Mint -> Unfreeze -> Transfer -> Freeze
Associate -> Mint -> Transfer
Associate -> Transfer
```
## Backwards Compatibility

> ## Security Implications

Reviewer 3: Q: Are there cases where multiple operators are involved or will the batch transactions always see a single entity carrying out some steps?
If there's a case where the batch contains entity A doing one thing and entity B doing something else there would be potential security considerations to address

Author: Yes, the transaction can involve multiple parties, but every individual transaction in the batch needs to fulfill the standard transaction signing requirements, so I don't see any added security implications.

Reviewer 7: @Nana-EC are you referring to the client operator here?

## How to Teach This

## Reference Implementation

## Rejected Ideas

> We rejected the idea to add the support of conditional branching to the batch transactions,

Reviewer 3: Can you expand on this. Did you mean you rejected the case where essentially some transactions are carried out but not all?

Author: Yes, basically we won't support conditional branching, understood as if transaction B is not successful then execute transaction C.
